src/watermarks.ml

```diff
@@ -174,28 +177,53 @@ let get_name ~files ?name () =
       else
         None)
   in
-  if package_names = [] then die "@{<error>Error@}: no <package>.opam files found.";
-  match name with
-  | Some name ->
-    if not (List.mem name ~set:package_names) then
-      die "@{<error>Error@}: file %s.opam doesn't exist." name;
-    name
-  | None ->
-    let shortest =
-      match package_names with
-      | [] -> assert false
-      | first :: rest ->
-        List.fold_left rest ~init:first ~f:(fun acc s ->
-          if String.length s < String.length acc then
-            s
+  if package_names = [] then
+    die "@{<error>Error@}: no <package>.opam files found.";
+  let name =
+    match Which_program.t with
+    | Dune -> begin
+        assert (Option.is_none name);
+        if not (List.mem ~set:files Dune_project.filename) then
+          die "@{<error>Error@}: There is no dune-project file in the current \
+               directory, please add one with a (name <name>) field in it.\n\
+               Hint: dune subst must be executed from the root of the project.";
+        match read_project_name () with
+        | None ->
+          die "@{<error>Error@}: The project name is not defined, please add \
+               a (name <name>) field to your dune-project file."
+        | Some name -> name
+      end
+    | Jbuilder ->
+      match name with
+      | Some name -> name
+      | None ->
+        match
+          if List.mem ~set:files Dune_project.filename then
+            read_project_name ()
           else
-            acc)
-    in
-    if List.for_all package_names ~f:(String.is_prefix ~prefix:shortest) then
-      shortest
-    else
-      die "@{<error>Error@}: cannot determine name automatically.\n\
-           You must pass a [--name] command line argument."
+            None
+        with
+        | Some name -> name
+        | None ->
+          let shortest =
+            match package_names with
+            | [] -> assert false
+            | first :: rest ->
+              List.fold_left rest ~init:first ~f:(fun acc s ->
+                if String.length s < String.length acc then
+                  s
+                else
+                  acc)
+          in
+          if List.for_all package_names ~f:(String.is_prefix ~prefix:shortest) then
+            shortest
+          else
+            die "@{<error>Error@}: cannot determine name automatically.\n\
+                 You must pass a [--name] command line argument."
+  in
+  if not (List.mem name ~set:package_names) then
+    die "@{<error>Error@}: file %s.opam doesn't exist." name;
```

Comment:
Looking at this again, the error on line 180 is slightly different. So ignore that comment. Everything looks fine. Though it might still be useful to include `package_names` in the error for convenience.